README.md
# decision-tree

A Clojure library designed to run decision-tree. 
The library's decision-tree targets classification problem.

## Usage
* Add :dependencies in your project.clj file.
```
:dependencies [[org.clojure/clojure "1.10.1"]
               [mrcsce/decision-tree "0.1.0"]]
```

* Add :require in your clj file to import the library.
```
(:require [decision-tree.core :as dt])
```

* Sample code (Note: If you want to run this sample code, you also add `[incanter "1.5.5"]` in your dependencies.)
```clojure
(ns your-project.core
  (:require [decision-tree.core :as dt])
  (:require [incanter.datasets]))

;; To indicate sample code, we use iris dataset.
(def iris (->> (incanter.datasets/get-dataset :iris)
               :rows))

;; iris dataset is like a following data structure.
(take 3 iris)
;; =>({:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Species \"setosa\"}
;     {:Sepal.Length 4.9, :Sepal.Width 3.0, :Petal.Length 1.4, :Petal.Width 0.2, :Species \"setosa\"}
;;    {:Sepal.Length 4.7, :Sepal.Width 3.2, :Petal.Length 1.3, :Petal.Width 0.2, :Species \"setosa\"})

;; split iris-dataset to test-data and train-data.
(def test-data (random-sample 0.3 iris))
(def train-data (filter #(not (.contains test-data %1)) iris))

;; max-depth-of-decision-tree is parameter of decision-tree.
;; If you set too big value to this, more likely overfitting.
;; If you set too small value to this, more likely underfitting.
(def max-depth-of-decision-tree 3)

;; A signature of dt/make-decision-tree is `make-decision-tree [train-data max-depth key-of-objective-variable]`.
;; `train-data` is coll of map(data).
;; `max-depth` is positive integer.
;; `key-of-objective-variable` is keyword which we want to predict for example it is :Species in iris dataset.
(def decision-tree (dt/make-decision-tree train-data max-depth-of-decision-tree :Species))

;; Example of predict function in this library.
;; The predict function take just one data like a
;; {:Sepal.Length 5.9, :Sepal.Width 3.0, :Petal.Length 5.1, :Petal.Width 1.8, :Species \"virginica\"})
;; or {:Sepal.Length 5.9, :Sepal.Width 3.0, :Petal.Length 5.1, :Petal.Width 1.8}.
(dt/predict decision-tree (first test-data))
;; => \"setosa\"

;; the predict is correct.
(first test-data)
;; => {:Sepal.Length 5.0, :Sepal.Width 3.4, :Petal.Length 1.5, :Petal.Width 0.2, :Species \"setosa\"}\n


;; Example of correct-rate in all test-data.
(def results
  (map #(= %1 %2)
       (->> test-data
            (map (partial dt/predict decision-tree)))
       (->> test-data
            (map :Species))))

(defn calculate-correct-rate
  [results]
  (let [results (frequencies results)
        k (->> (keys results)
               (map str)
               (map keyword))
        v (vals results)
        r (zipmap k v)
        number-of-true (:true r)
        number-of-data (+ (:true r) (:false r))]
    (->> (/ number-of-true number-of-data)
         double)))

(calculate-correct-rate results)
;;=> 0.9761904761904762
;; Accomplish 97% accuracy in test data!!
```

## License

Copyright 2020 Ryota Miyoshi

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-08-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-08-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/decision-tree/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/decision-tree/compare/0.1.0...0.1.1


test/decision_tree/core_test.clj
(ns decision-tree.core-test
  (:require [clojure.test :refer :all]
            [decision-tree.core :refer :all]))

(deftest test-update-key
  (testing "expectation use case of the function"
    (let [input-data {:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Species "setosa"}
          expected-output-data {:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Classes "setosa"}]
      (is (= expected-output-data
             (#'decision-tree.core/update-key :Species :Classes input-data))))))

(deftest test-gini-impurity
  (testing "expectation use case of the function"
    (let [data-expected-zero '("a" "a" "a")
          data-expected-two-thirds '("a" "b" "c")]
      (is (= 0.0 (#'decision-tree.core/gini-impurity data-expected-zero)))
      (is (= (double (/ 2 3)) (#'decision-tree.core/gini-impurity data-expected-two-thirds))))))


src/decision_tree/core.clj
(ns decision-tree.core
  (:require [clojure.spec.alpha :as s]))


(defn- update-key
  "Change `from-key` in `map` to `to-key`.
  Example:
    (def sample-map {:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Species \"setosa\"})
    (update-key :Species :Classes sample-map)
    ;;=> {:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Classes \"setosa\"}"
  [from-key to-key map]
  {:pre [(s/valid? keyword? from-key)
         (s/valid? keyword? to-key)
         (s/valid? map? map)]
   :post [(s/valid? map? %)]}
  (-> map
      (assoc to-key (from-key map))
      (dissoc from-key)))

(defn- update-value
  "Update value related to `key` in `map`.
  Example:
    (update-value :test 10 {:test nil :aiueo 5})
    ;;=> {:test 10 :aiueo 5}
  "
  [map key updated-value]
  {:pre [(s/valid? map? map)
         (s/valid? keyword? key)]
   :post [(s/valid? map? %)]}
  (update map key (fn [_] (identity updated-value))))


(s/def ::objective-variable string?)
(s/def ::objective-variable-vector (s/coll-of ::objective-variable))

(defn- gini-impurity
  "Calculate gini-impurity.
  `y` is objective variable vector."
  [y]
  {:pre [(s/valid? ::objective-variable-vector y)]
   :post [(s/valid? (s/and number? #(<= 0 % 1)) %)]}
  (let [number-of-each-data (->> (group-by identity y)
                                 (map (comp count val)))
        sum-of-number-of-data (apply + number-of-each-data)]
    (->> number-of-each-data
         (map #(/ % sum-of-number-of-data))
         (map #(Math/pow % 2))
         (apply +)
         (- 1))))

(defn information-gain [node-data leaf1-data leaf2-data]
  (let [node-number-of-data (count node-data)
        node-gini-impurity (gini-impurity (map :Classes node-data))
        leaf1-number-of-data (count leaf1-data)
        leaf1-gini-impurity (gini-impurity (map :Classes leaf1-data))
        leaf2-number-of-data (count leaf2-data)
        leaf2-gini-impurity (gini-impurity (map :Classes leaf2-data))]
    (- node-gini-impurity
       (+ (* (/ leaf1-number-of-data node-number-of-data) leaf1-gini-impurity)
          (* (/ leaf2-number-of-data node-number-of-data) leaf2-gini-impurity)))))


(s/def ::feature (s/or :nil nil?
                       :keyword keyword?))
(s/def ::threshold (s/or :nil nil?
                         :number number?))
(s/def ::data (s/coll-of map?))
(s/def ::left (s/or :nil nil?
                    :map? ::node))
(s/def ::right (s/or :nil nil?
                     :map? ::node))
(s/def ::node  (s/keys :req-un [::feature ::threshold ::data ::left ::right]))

(defn- count-number-of-kinds-of-objective-variables
  "Example:
    (count-number-of-kinds-of-objective-variables iris :Classes)
    ;;=> 3"
  [data key-of-objective-variable]
  {:pre [(s/valid? ::data data)
         (s/valid? keyword? key-of-objective-variable)]
   :post [(s/valid? int? %)]}
  (->> data
       (map key-of-objective-variable)
       set
       count))

(defn- get-explanatory-variables-from
  "Example:
    (get-explanatory-variables-from iris :Classes)
    ;;=> (:Sepal.Length :Sepal.Width :Petal.Length :Petal.Width)"
  [data key-of-objective-variable]
  {:pre [(s/valid? ::data data)
         (s/valid? keyword? key-of-objective-variable)]
   :post [s/valid? (s/coll-of keyword?) %]}
  (->> (first data)
       keys
       (filter (partial not= key-of-objective-variable))))


(defn- get-threshold-point-candidates
  "Get candidates of threshold to split node to right node and left node.

  Now implementation of this function is that the function takes data
  then returns coll of map of explanatory-variable and the value.
  Example:
    (take 3 iris)
    ;;=>({:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Classes \"setosa\"}
         {:Sepal.Length 4.9, :Sepal.Width 3.0, :Petal.Length 1.4, :Petal.Width 0.2, :Classes \"setosa\"}
         {:Sepal.Length 4.7, :Sepal.Width 3.2, :Petal.Length 1.3, :Petal.Width 0.2, :Classes \"setosa\"})
    (get-threshold-point-candidates (take 3 iris))
    ;;=> ({:Sepal.Length 5.1}
          {:Sepal.Length 4.9}
          {:Sepal.Length 4.7}
          {:Sepal.Width 3.5}
          {:Sepal.Width 3.0}
          {:Sepal.Width 3.2}
          {:Petal.Length 1.4}
          {:Petal.Length 1.4}
          {:Petal.Length 1.3}
          {:Petal.Width 0.2}
          {:Petal.Width 0.2}
          {:Petal.Width 0.2})"
  [data]
  {:pre [(s/valid? ::data data)]
   :post [(s/valid? (s/coll-of map?) %)]}
  (let [features (get-explanatory-variables-from data :Classes)
        get-threshold-point-candidates-one-feature (fn [data feature] (map #(select-keys %1 [feature]) data))]
    (->> (map (partial get-threshold-point-candidates-one-feature data) features)
         flatten)))

(defn- create-node-from-data
  [data]
  {:pre [(s/valid? ::data data)]
   :post [s/valid? ::node %]}
  {:feature nil :threshold nil :data data :right nil :left nil})

(defn- split-one-node
  [node threshold key]
  {:pre [(s/valid? ::node node)
         (s/valid? ::threshold threshold)
         (s/valid? ::feature key)]
   :post [(s/valid? ::node %)]}
  (let [left-data (filter #(> (key %1) threshold) (:data node))
        left-node (create-node-from-data left-data)
        right-data (filter #(<= (key %1) threshold) (:data node))
        right-node (create-node-from-data right-data)]
    (-> node
        (update-value :feature key)
        (update-value :threshold threshold)
        (update-value :left left-node)
        (update-value :right right-node))))

(defn- calculate-information-gains
  "Calculate information-gain as all split pattern.
  Example:
    (take 3 (calculate-information-gains (create-node-from-data iris)))
    ;;=> ({:information-gain 0.1684194823599613, :threshold 5.1, :feature :Sepal.Length}
          {:information-gain 0.08546401515151525, :threshold 4.9, :feature :Sepal.Length}
          {:information-gain 0.052757793764988015, :threshold 4.7, :feature :Sepal.Length})"
  [node]
  {:pre  [(s/valid? ::node node)]
   :post [(s/valid? (s/coll-of map?) %)]}
  (->> (get-threshold-point-candidates (:data node))
       (map #(split-one-node node ((comp first vals) %1) ((comp first keys) %1)))
       (map #(hash-map
               :information-gain (information-gain (:data %) (:data (:left %)) (:data (:right %)))
               :threshold (:threshold %)
               :feature (:feature %)))))



(defn- stop-split?
  "Judge whether recursion is stop ot not. recursion is stop when
  max-depth is 0 or
  number of kinds of objective-variables is 1 or 0.
  number of kinds of objective-variables is 0 means that data is empty"
  [node max-depth]
  (or (= max-depth 0)
      (<= (count (:data node)) 1)
      (<= (count-number-of-kinds-of-objective-variables (:data node) :Classes) 1)))

(defn- get-maximum-information-gain-splitter
  "Example:
    (get-maximum-information-gain-splitter (create-node-from-data iris))
    ;;=> {:information-gain 0.33333333333333337, :threshold 0.6, :feature :Petal.Width)"
  [node]
  (->> (calculate-information-gains node)
       ;; max-key returns a last element if there are exact same values.
       ;; So we added shuffle because we want this function returns a random element in exact same values.
       shuffle
       (apply max-key :information-gain)))

(defn- get-most-popular-objective-variable-values
  [data]
  (->> data
       (map :Classes)
       (frequencies)
       (apply max-key val)
       key))

(defn- split-node
  "Split node to make decision-tree."
  [node threshold key max-depth]
  {:pre [(s/valid? ::node node)
         (s/valid? ::threshold threshold)
         (s/valid? ::feature key)]
   :post [(s/valid? ::node %)]}
  (if (stop-split? node max-depth)
    (assoc node :predict (get-most-popular-objective-variable-values (:data node)))
    (let [split (split-one-node node threshold key)
          left-node (:left split)
          right-node (:right split)
          left-splitter (get-maximum-information-gain-splitter left-node)
          right-splitter (get-maximum-information-gain-splitter right-node)]
      (-> node
          (update-value :feature key)
          (update-value :threshold threshold)
          (update-value :left (split-node left-node (:threshold left-splitter) (:feature left-splitter) (dec max-depth)))
          (update-value :right (split-node right-node (:threshold right-splitter) (:feature right-splitter) (dec max-depth)))))))

(defn make-decision-tree
  "Make decision-tree.
  `train-data` is coll of map.
  `max-depth` is max-depth of decision-tree made
  `key-of-objective-variable` is key-of-objective-variable for example if use iris dataset this is :Species
  Example of `make-decision-tree` used iris dataset :
    (take 3 iris)
    ;; =>({:Sepal.Length 5.1, :Sepal.Width 3.5, :Petal.Length 1.4, :Petal.Width 0.2, :Species \"setosa\"}
          {:Sepal.Length 4.9, :Sepal.Width 3.0, :Petal.Length 1.4, :Petal.Width 0.2, :Species \"setosa\"}
          {:Sepal.Length 4.7, :Sepal.Width 3.2, :Petal.Length 1.3, :Petal.Width 0.2, :Species \"setosa\"}
    (def decision-tree (make-decision-tree iris 3 :Species)
  "
  [train-data max-depth key-of-objective-variable]
  {:pre [(s/valid? ::data train-data)]
   :post [(s/valid? ::node %)]}
  (let [train-data (map (partial update-key key-of-objective-variable :Classes) train-data)
        node (create-node-from-data train-data)
        splitter (get-maximum-information-gain-splitter node)]
    (split-node node (:threshold splitter) (:feature splitter) max-depth)))

(defn predict
  "Predict objective-variable used by decision-tree
  `tree` is decision-tree made by `make-decision-tree` function.
  `data` is just one data like a
  Example of `data`:
    {:Sepal.Length 6.2, :Sepal.Width 3.4, :Petal.Length 5.4, :Petal.Width 2.3, :Species \"virginica\"}"
  [tree data]
  (cond (not (nil? (:predict tree)))
        (:predict tree)
        (> ((:feature tree) data) (:threshold tree)) (predict (:left tree) data)
        :else (predict (:right tree) data)))
